File: src/codes/tight-coupling/py/rfc_di.py
from abc import ABC, abstractmethod

class Notifier(ABC):
    @abstractmethod
    def send(self, email: str, message: str):
        pass

class Mailer(Notifier):
    def send(self, email: str, message: str):
        print(f"送信しました: {email} → {message}")

class UserService:
    def __init__(self, notifier: Notifier):
        self.notifier = notifier

    def register_user(self, email: str):
        print(f"ユーザー登録: {email}")
        self.notifier.send(email, "ようこそ！")

# 利用例
mailer = Mailer()
service = UserService(mailer)
service.register_user("user@example.com")

File: src/codes/tight-coupling/ts/rfc_di.ts
// インターフェースで依存先を抽象化（任意）
interface Notifier {
  send(email: string, message: string): void;
}

// 実装クラス
class Mailer implements Notifier {
  send(email: string, message: string): void {
    console.log(`送信しました: ${email} → ${message}`);
  }
}

// クライアントクラス
class UserService {
  constructor(private notifier: Notifier) {}

  registerUser(email: string): void {
    console.log(`ユーザー登録: ${email}`);
    this.notifier.send(email, "ようこそ！");
  }
}

// 利用例（依存を注入）
const mailer = new Mailer();
const service = new UserService(mailer);
service.registerUser("user@example.com");
